thirdparty/github.go

```diff
@@ -1919,3 +1923,138 @@ func GetBranchProtectionRules(ctx context.Context, token, owner, repo, branch st
 	}
 	return nil, nil
 }
+
+// getGithubAppAuth returns the app id and app private key if they exist.
+func getGithubAppAuth(s *evergreen.Settings) *githubAppAuth {
+	if s.AuthConfig.Github == nil || s.AuthConfig.Github.AppId == 0 {
+		return nil
+	}
+
+	key := s.Expansions[evergreen.GithubAppPrivateKey]
+	if key == "" {
+		return nil
+	}
+
+	return &githubAppAuth{
+		appId:      s.AuthConfig.Github.AppId,
+		privateKey: []byte(key),
+	}
+}
+
+// CreateInstallationToken uses the owner/repo information to request an github app installation id
+// and uses that id to create an installation token.
+func CreateInstallationToken(ctx context.Context, s *evergreen.Settings, owner, repo string, opts *github.InstallationTokenOptions) (string, error) {
+	const (
+		maxDelay   = 10 * time.Second
+		minDelay   = time.Second
+		maxRetries = 5
+	)
+
+	if owner == "" || repo == "" {
+		return "", errors.New("no owner/repo specified to create installation token")
+	}
+	authFields := getGithubAppAuth(s)
+	if authFields == nil {
+		return "", errors.New("GitHub app is not configured in admin settings")
+	}
+
+	retryConf := utility.NewDefaultHTTPRetryConf()
+	retryConf.MaxDelay = maxDelay
+	retryConf.BaseDelay = minDelay
+	retryConf.MaxRetries = maxRetries
+
+	httpClient := utility.GetHTTPRetryableClient(retryConf)
+	defer utility.PutHTTPClient(httpClient)
+
+	key, err := jwt.ParseRSAPrivateKeyFromPEM(authFields.privateKey)
+	if err != nil {
+		return "", errors.Wrap(err, "parsing private key")
+	}
+
+	itr := ghinstallation.NewAppsTransportFromPrivateKey(httpClient.Transport, authFields.appId, key)
+	httpClient.Transport = itr
+	client := github.NewClient(httpClient)
+
+	installationID, err := dbGithub.GetInstallationID(ctx, owner, repo)
+	if err != nil {
+		return "", errors.Wrapf(err, "getting cached installation id for '%s/%s'", owner, repo)
+	}
+	if installationID == 0 {
+		installation, resp, err := client.Apps.FindRepositoryInstallation(ctx, owner, repo)
```

Comment:
ideally thirdparty functions just have the logic of calling Github, so a FindInstallation function and a CreateInstallationToken function, and then some other function could do this additional logic of checking the cache, updating the cache, calling those thirdparty functions (like this function for example: https://github.com/evergreen-ci/evergreen/blob/ff46048d63feb595caedb68fd4d9fc85ab73a96b/model/patch/patch.go#L1038  -- in this case, it seems like that could live in the new model/github/installation.go library)